pages/blog/guide-to-nova-gc.md

```diff
@@ -0,0 +1,678 @@
+and `Object`.) Normally, in Rust you see lifetimes mostly in references. For
+reasons explained above, Nova cannot use the `Agent` reference to derive the
+lifetime for bindable values. Nova also cannot quite use normal Rust lifetime
+rules directly due to Rust not allowing aliasing. For this reason, bindable
+values implement the following trait:
+
+```rust
+unsafe trait Bindable {
+    type Of<'a>;
+
+    fn unbind(self) -> Self::Of<'static>;
+    fn bind<'a>(self, gc: NoGcScope<'a, '_>) -> Self::Of<'a>;
+}
+```
+
+The `Of` type is expected to (and once possible to check, be required to) be
+equal to the `Self` type. For example, this is how `Value`'s implementation
+begins:
+
+```rust
+unsafe impl Bindable for Value<'_> {
+    type Of<'a> = Value<'a>;
+}
+```
+
+What this says, effectively, is that for any `Value`, regardless of what
+lifetime it has, it has a function `unbind` that takes self (by value, not by
+reference) and returns a `Value<'static>`, and a `bind` function that again
+takes takes self and a `NoGcScope<'a, '_>` and returns a `Value<'a>`. You can
+perhaps see where this is going: "Bound values" mean values that carry the `'a`
+lifetime from a `NoGcScope<'a, '_>`, either from having had `bind(gc.nogc())`
+called on them, or from being returned from a function that takes a
+`NoGcScope<'a, '_>` and returns the value with the `'a` lifetime.
+
+It is also possible to get "exclusively bound values"; these are created when a
+bound value is returned with the `'a` lifetime from a function that takes
+`GcScope<'a, '_>`, ie. when a bound value is returned from a function that can
+trigger garbage collection. Due to certain details of Rust's borrow checker
+rules related to internal mutation, the returned value will keep the `GcScope`
+from being reused. We'll see how to deal with this situation later.
+
+## The Practice
+
+Now that we have seen the `GcScope`, `NoGcScope`, and bound values, we are ready
+to start using them in practice. First, let's do a quick review of these three
+principal actors:
+
+1. `GcScope<'a, '_>`: A zero-sized type that is passed to functions that can
+   trigger garbage collection. Only one such type is ever active at a time, and
+   no `NoGcScope`s can be active at the same time.
+2. `NoGcScope<'a, '_>`: A zero-sized type that is passed to functions that
+   cannot trigger garbage collection. Any number of such types can be active at
+   the same time, but no `GcScope` can be active at the same time.
+3. Bound values: Handles to garbage collected values that have been bound to a
+   `NoGcScope` or `GcScope`, implicitly or explicitly.
+
+With this, we're ready to start our praxis.
+
+### Returning bound values from `NoGcScope` functions
+
+The simplest function that deals with bound values we can imagine in the engine
+is one that cannot trigger garbage collection, and returns a bound value. An
+example of this would be `Value::from_string` which takes a Rust
```

Comment:
I'm not sure that `Value::from_string` is a good example, because this is a function that *should GC* if the heap vector is full (enough). In fact, I think most of the functions that take a `NoGcScope` and return a value currently in Nova should take a `GcScope` instead because of this.

Reply:
Mmm... I wonder what would be a better choice.. `Value::from_bool` would work if it existed, I guess, but on the other hand that's also guaranteed to never allocate anything and the `Value` returned from such a function would be perfectly valid to return as `Value<'static>` instead...

I guess eventually, if we do allow GC in heap allocating functions, then we'd also have non-GC allocating functions that would panic if there is no space for the allocation. Those would then be good candidates for this example, but for now I cannot really think of a non-allocating method that takes some plain-old-data parameter and returns a bound value.

Reply:
In which cases would you use a non-GC allocating function? I think the only cases where you can be sure the function wouldn't panic is immediately after a GC, and most functions in most of the engine can't actually know if the last `gc.reborrow()` call actually GC'd.

Reply:
Hot loops: Check that there is space on the heap, reserve if not, and then perform non-GC allocation inside the loop.